nexus/src/app/background/blueprint_load.rs

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for loading the target blueprint from the DB
+//!
+//! This task triggers the `plan_execution` background task when the blueprint
+//! changes.
+
+use super::common::BackgroundTask;
+use futures::future::BoxFuture;
+use futures::FutureExt;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use nexus_types::deployment::Blueprint;
+use serde_json::json;
+use std::sync::Arc;
+use tokio::sync::watch;
+
+pub struct TargetBlueprintLoader {
+    datastore: Arc<DataStore>,
+    last: Option<Arc<Blueprint>>,
+    tx: watch::Sender<Option<Arc<Blueprint>>>,
+    rx: watch::Receiver<Option<Arc<Blueprint>>>,
+}
+
+impl TargetBlueprintLoader {
+    pub fn new(datastore: Arc<DataStore>) -> TargetBlueprintLoader {
+        let (tx, rx) = watch::channel(None);
+        TargetBlueprintLoader { datastore, last: None, tx, rx }
+    }
+
+    /// Expose the target blueprint
+    pub fn watcher(&self) -> watch::Receiver<Option<Arc<Blueprint>>> {
+        self.rx.clone()
+    }
+}
+
+impl BackgroundTask for TargetBlueprintLoader {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        async {
+            // Set up a logger for this activation that includes metadata about
+            // the current target.
+            let log = match &self.last {
+                None => opctx.log.clone(),
+                Some(old) => opctx.log.new(o!(
+                    "current_target_id" => old.id.to_string(),
+                    "current_time_created" => old.time_created.to_string(),
+                )),
+            };
+
+            // Retrieve the latest target blueprint
+            let result =
+                self.datastore.blueprint_target_get_current_full(opctx).await;
+
+            // Decide what to do with the result
+            match (&self.last, result) {
+                (_, Err(error)) => {
+                    // We failed to read the blueprint. There's nothing to do
+                    // but log an error. We'll retry when we're activated again.
+                    let message = format!("{:#}", error);
+                    warn!(
+                        &log,
+                        "failed to read target blueprint";
+                        "error" => &message
+                    );
+                    json!({
+                        "error": 
+                        format!("failed to read target blueprint: {message}")})
```

Comment:
nit: the formatting here looks a little weird.

Reply:
fixed in [596c9be](https://github.com/oxidecomputer/omicron/pull/4891/commits/596c9be466c0ba54c41201edf0c18d9b8c6c6280)